### PR TITLE
[MIG][13.0] sale_order_general_discount

### PR DESCRIPTION
**Pull request by acsonefho**
Migration in v13.0 of `sale_order_general_discount` module.

The work was already done by this PR: https://github.com/OCA/sale-workflow/pull/981

But not possible to create a MR to include it.

### CODE REVIEW

**Comment by rousseldenis** [comment]
@acsonefho Is the PR #981 abandonned ?

**Comment by acsonefho** [comment]
> @acsonefho Is the PR #981 abandonned ?

@rousseldenis I suppose, no more action for 3 months and the PR still red.

File: sale_order_general_discount/models/sale_order.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        digits="Discount",
+    )
+
+    @api.depends("partner_id")
+    def _compute_general_discount(self):
+        for so in self:
+            so.general_discount = so.partner_id.sale_discount
+
+    @api.onchange("general_discount")
+    def onchange_general_discount(self):
```
Comment (sergio-teruel) [inline sale_order_general_discount/models/sale_order.py]: Remove onchange method, this not needed

File: sale_order_general_discount/models/sale_order_line.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    discount = fields.Float(
+        compute="_compute_general_discount",
+        store=True,
+        readonly=False,
+        string="Discount (%)",
+        digits="Discount",
+        default=0.0,
+    )
+
+    @api.depends("order_id")
```
Comment (sergio-teruel) [inline sale_order_general_discount/models/sale_order_line.py]:
```suggestion
    @api.depends("order_id.general_discount")
```

File: sale_order_general_discount/tests/test_sale_order_general_discount.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            }
+        )
+        cls.View = cls.env["ir.ui.view"]
+
+    def test_default_partner_discount(self):
+        self.assertEqual(self.order.general_discount, self.partner.sale_discount)
+
+    def test_sale_order_values(self):
+        self.order.general_discount = 10
+        self.order.onchange_general_discount()
```
Comment (sergio-teruel) [inline sale_order_general_discount/tests/test_sale_order_general_discount.py]: Remove onchange method from test

**Review by sergio-teruel — CHANGES_REQUESTED** [review]
Minor changes, and the most important, remove onchange method

File: sale_order_general_discount/models/sale_order_line.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2018 Tecnativa - Sergio Teruel
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    discount = fields.Float(
+        compute="_compute_general_discount",
```
Comment (pedrobaeza) [inline sale_order_general_discount/models/sale_order_line.py]: Use convention for name and for not having problems with other inheritances:
```suggestion
        compute="_compute_discount",
```

File: sale_order_general_discount/models/sale_order_line.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        store=True,
+        readonly=False,
+        string="Discount (%)",
+        digits="Discount",
+        default=0.0,
```
Comment (pedrobaeza) [inline sale_order_general_discount/models/sale_order_line.py]: Please don't redefine these already set attributes. It forces you to re-translate without need (or to be inconsistent with upstream).
```suggestion
```

File: sale_order_general_discount/models/sale_order_line.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    )
+
+    @api.depends("order_id")
+    def _compute_general_discount(self):
+        for line in self:
```
Comment (pedrobaeza) [inline sale_order_general_discount/models/sale_order_line.py:15]: For being friendly with inheritance, you must put:
```suggestion
        it hasattr(super(), "_compute_discount"):
            super()._compute_discount()
        for line in self:
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by acsonefho** [comment]
@pedrobaeza @sergio-teruel Changes done!
Thank you for your review!

**Review by sergio-teruel — APPROVED** [review]

**Review by pedrobaeza — COMMENTED** [review]
Is the `fields_view_get` with the context of the default discount still needed with the new computed writeable field?

Please squash anyway migration commits together (yours and the other from Sudhir) for having a bit of organized timeline.

**Comment by sergio-teruel** [comment]
Is needed only for user can be see the discount value Immediately, otherwise, user see discount value after saves the record

**Review by HaraldPanten — CHANGES_REQUESTED** [review]
After functional testing:

0- Activate the settings that you need.
1- Create a quotation with several lines.
2- Set a Global discount.
3- Save or validate.

Result: Discount is not being applied.

Could you check?

**Comment by pedrobaeza** [comment]
> Is needed only for user can be see the discount value Immediately, otherwise, user see discount value after saves the record

OK, then only squashing is left for merging this.

**Comment by sergio-teruel** [comment]
@HaraldPanten Are you right... is due to use the compute method with other modules...

**Review by sergio-teruel — CHANGES_REQUESTED** [review]
This module with last changes it's failing

**Comment by pedrobaeza** [comment]
The problem seems to be due to other modules in the repo hijacking the method name. I have fixed it in 841e7aa1f5a6e438f7af95c27a64933bb58b4fa8.

Can you please rebase (and do the squashing)?

**Comment by sergio-teruel** [comment]
@acsonefho CAn you attend lastest comments?? Thanks...

**Review by sergio-teruel — APPROVED** [review]

**Comment by sergio-teruel** [comment]
@pedrobaeza @HaraldPanten Can you review??

**Comment by pedrobaeza** [comment]
Please squash all migration commits into one.

**Review by HaraldPanten — APPROVED** [review]
> After functional testing:
> 
> 0- Activate the settings that you need.
> 1- Create a quotation with several lines.
> 2- Set a Global discount.
> 3- Save or validate.
> 
> Result: Discount is not being applied.
> 
> Could you check?

Now it has been solved and IMO it's functionally working fine. Just squash your commits in the relevant ones.

**Comment by acsonefho** [comment]
> Please squash all migration commits into one.



> > After functional testing:
> > 0- Activate the settings that you need.
> > 1- Create a quotation with several lines.
> > 2- Set a Global discount.
> > 3- Save or validate.
> > Result: Discount is not being applied.
> > Could you check?
> 
> Now it has been solved and IMO it's functionally working fine. Just squash your commits in the relevant ones.

@pedrobaeza @HaraldPanten Like this or I should `squash` more? As we should split `pre-commit` commit and the migration. But the migration part has be done in the original PR and I just did minor modifications.

**Comment by pedrobaeza** [comment]
Not really, but there's an initial problem:

Selección_026

Previous migration was not done properly, splitting the initial pre-commit pass, so you are dragging that problem. At least that commits should be squashed, but my favorite option is that you get rid off the previous migration, and start the proper migration procedure from scratch, and then mount your final code in top of it.

**Comment by rousseldenis** [comment]
@acsonefho Could you attend @pedrobaeza comments?

**Comment by Rad0van** [comment]
Works fine for me. If it only requires re-doing the migration and @acsonefho doesn't have time for this I could volunteer.

**Comment by maitaoriana** [comment]
Hi @pedrobaeza, I made the changes requested by you on this PR #1281. This PR have the commits made by @sudhir-erpharbor and @acsonefho but with squash and pre-commit . Please, Could you review it? Thanks.
Regards

**Comment by acsonefho** [comment]
I close this PR in favor of https://github.com/OCA/sale-workflow/pull/1281

**Pull request closed without merging**